File: SongsWebApp/SongsDatabase/DataModels/AlbumPlayed.cs
namespace SongsDatabase.DataModels
{
    using System;

    using MongoDB.Bson.Serialization.Attributes;

    using Base;


    [BsonIgnoreExtraElements]
    public class AlbumPlayed : BaseMongoEntity
    {
        #region Properties

        /// <summary>
        /// Gets or sets the date that the album was listened on.
        /// </summary>
        [BsonElement("date")]
        public DateTime Date { get; set; }

        /// <summary>
        /// Gets or sets the location that the album was listened in.
        /// </summary>
        [BsonElement("location")]
        public string Location { get; set; }

        /// <summary>
        /// Gets or sets the artist played.
        /// </summary>
        [BsonElement("artist")]
        public string Artist { get; set; }

        /// <summary>
        /// Gets or sets the album played.
        /// </summary>
        [BsonElement("album")]
        public string Album { get; set; }

        /// <summary>
        /// Gets or sets the user.
        /// </summary>
        [BsonElement("user_name")]
        public string UserName { get; set; }

        /// <summary>
        /// Gets or sets the web address for the album cover image.
        /// </summary>
        [BsonElement("image_path")]
        public string ImagePath { get; set; }

        /// <summary>
        /// Gets or sets the link to play the album.
        /// </summary>
        [BsonElement("player_link")]
        public string PlayerLink { get; set; }

        #endregion

        #region BaseMongoEntity Implementation

        public override string EquivalenceName => Date.ToShortDateString() + " " +
                                                  Location + " " +
                                                  Artist + " " +
                                                  Album + " " +
                                                  UserName;

        #endregion

        #region  Constructors

        public AlbumPlayed()
        {
            Date = DateTime.Now;
            Location = string.Empty;
            Artist = string.Empty;
            Album = string.Empty;
            UserName = string.Empty;
            ImagePath = string.Empty;
            PlayerLink = string.Empty;
        }

        public AlbumPlayed(AlbumPlayed album)
        {
            Date = album.Date;
            Location = album.Location;
            Artist = album.Artist;
            Album = album.Album;
            UserName = album.UserName;
            ImagePath = album.ImagePath;
            PlayerLink = album.PlayerLink;
        }

        #endregion 
    }
}

File: SongsWebApp/SongsBurger/ClientApp/app/Models/AlbumPlayed.ts
﻿import { IDbElement } from './IDbElement';

export interface IAlbumPlayed extends IDbElement
{
    date: Date;
    location: string;
    artist: string;
    album: string;
    userName: string;
    imagePath: string;
    playerLink: string;
};


export class AlbumPlayed implements IAlbumPlayed
{
    static fromData(data: IAlbumPlayed)
    {
        return new this(
            data.id,
            data.date,
            data.location,
            data.artist,
            data.album,
            data.userName,
            data.imagePath,
            data.playerLink);
    }

    constructor(
        public id: string = "",
        public date: Date = new Date(),
        public location: string = "",
        public artist: string = "",
        public album: string = "",
        public userName: string = "",
        public imagePath: string = "",
        public playerLink: string = "")
    {
    }
};


export interface IAlbumPlayedAddResponse
{
    album: IAlbumPlayed;
    errorCode: number;
    failReason: string;
    userId: string;
};

export class AlbumPlayedAddResponse implements IAlbumPlayedAddResponse
{
    static fromData(data: IAlbumPlayedAddResponse)
    {
        var album: AlbumPlayed = AlbumPlayed.fromData(data.album);
        return new this(
            album,
            data.errorCode,
            data.failReason,
            data.userId);
    }

    constructor(
        public album: IAlbumPlayed = new AlbumPlayed(),
        public errorCode: number = -1,
        public failReason: string = "",
        public userId: string = "") {
    }
};

